contracts/tokenbridge/ethereum/gateway/L1FeeTokenUSDCCustomGateway.sol

```diff
@@ -0,0 +1,107 @@
+ *         - owner can trigger burning all the USDC tokens locked in the gateway
+ *
+ *         This contract is to be used on chains where custom fee token is used. If chain is using
+ *         ETH as native token then use L1USDCCustomGateway instead.
+ */
+contract L1FeeTokenUSDCCustomGateway is L1USDCCustomGateway {
+    using SafeERC20 for IERC20;
+
+    function _parseUserEncodedData(bytes memory data)
+        internal
+        pure
+        override
+        returns (uint256 maxSubmissionCost, bytes memory callHookData, uint256 tokenTotalFeeAmount)
+    {
+        (maxSubmissionCost, callHookData, tokenTotalFeeAmount) =
+            abi.decode(data, (uint256, bytes, uint256));
+    }
+
+    function _initiateDeposit(
+        address _refundTo,
+        address _from,
+        uint256, // _amount, this info is already contained in _data
+        uint256 _maxGas,
+        uint256 _gasPriceBid,
+        uint256 _maxSubmissionCost,
+        uint256 tokenTotalFeeAmount,
+        bytes memory _data
+    ) internal override returns (uint256) {
+        return sendTxToL2CustomRefund(
+            inbox,
+            counterpartGateway,
+            _refundTo,
+            _from,
+            tokenTotalFeeAmount,
+            0,
+            L2GasParams({
+                _maxSubmissionCost: _maxSubmissionCost,
+                _maxGas: _maxGas,
+                _gasPriceBid: _gasPriceBid
+            }),
+            _data
+        );
+    }
+
+    function _createRetryable(
+        address _inbox,
+        address _to,
+        address _refundTo,
+        address _user,
+        uint256 _totalFeeAmount,
+        uint256 _l2CallValue,
+        uint256 _maxSubmissionCost,
+        uint256 _maxGas,
+        uint256 _gasPriceBid,
+        bytes memory _data
+    ) internal override returns (uint256) {
+        {
+            // Transfer native token amount needed to pay for retryable fees to the inbox.
+            // Fee tokens will be transferred from user who initiated the action - that's `_user` account in
+            // case call was routed by router, or msg.sender in case gateway's entrypoint was called directly.
+            address nativeFeeToken = IERC20Bridge(address(getBridge(_inbox))).nativeToken();
+            uint256 inboxNativeTokenBalance = IERC20(nativeFeeToken).balanceOf(_inbox);
+            if (inboxNativeTokenBalance < _totalFeeAmount) {
+                address transferFrom = isRouter(msg.sender) ? _user : msg.sender;
+                IERC20(nativeFeeToken).safeTransferFrom(
+                    transferFrom, _inbox, _totalFeeAmount - inboxNativeTokenBalance
+                );
+            }
+        }
+
+        return IERC20Inbox(_inbox).createRetryableTicket(
+            _to,
+            _l2CallValue,
+            _maxSubmissionCost,
+            _refundTo,
+            _user,
+            _maxGas,
+            _gasPriceBid,
+            _totalFeeAmount,
+            _data
+        );
+    }
```

Comment:
this is now implemented multiple times in different Orbit version contracts, we should consider refactoring the code to use this as a library or base contract.